src/routes/hotels.php

```diff
@@ -12,6 +12,155 @@
+              GROUP BY h.id
+              HAVING COUNT(*) >= :rooms";
+
+    $db = new db();
+    $db = $db->connect();
+
+    try {
+        $stmt = $db->prepare($availabilityQueryTemplate);
+
+        $stmt->bindParam('checkin_date', $params['checkin_date']);
+        $stmt->bindParam('checkout_date', $params['checkout_date']);
+        $stmt->bindParam('rooms', $params['rooms']);
+
+        $stmt->execute();
+
+        $hotels = $stmt->fetchAll(PDO::FETCH_OBJ);
+
+        $db = null;
+
+        return $response->write(json_encode($hotels))->withStatus(200);
+    } catch (PDOException $e) {
+        $db = null;
+
+        $errorMessage = $e->getMessage();
+        return $response->write(json_encode(['error' => '' . $errorMessage]))->withStatus(500);
+    }
+});
+
+/**
+ * Return all the available rooms of a hotel
+ * Example GET params
+ *      checkin_date:2018-11-11
+ *      checkout_date:2019-01-02
+ * Example response
+ * [
+ * {
+ * "id": "20",
+ * "hotel_id": "1",
+ * "room_type_id": "36"
+ * }
+ * ]
+ * Note: at this point, we will no longer care about the number of rooms because the hotel you request for should already
+ * meet the requirement of having enough rooms for the request.
+ */
+$app->get('/api/hotels/{id}/availabilities', function (Request $request, Response $response, array $args) {
+
+    $params = $request->getQueryParams();
+
+    $hotelAvailability = "SELECT r.* from hotel h, room r, room_type rt 
+                          WHERE
+                              h.id = :id AND
+                              r.hotel_id = h.id AND
+                              r.room_type_id = rt.id AND 
+                              r.id NOT IN 
+                                (SELECT DISTINCT rr.room_id FROM reservation_room rr 
+                                WHERE 
+                                   (rr.checkout_date > :checkin_date AND rr.checkout_date < :checkout_date) OR 
+                                   (rr.checkin_date > :checkin_date AND rr.checkin_date < :checkout_date) OR 
+                                   (rr.checkin_date < checkin_date AND rr.checkout_date > :checkout_date))";
+
+    $db = new db();
+    $db = $db->connect();
+
+    try {
+        $stmt = $db->prepare($hotelAvailability);
+
+        $stmt->bindParam('id', $args['id']);
+        $stmt->bindParam('checkin_date', $params['checkin_date']);
+        $stmt->bindParam('checkout_date', $params['checkout_date']);
+
+        $stmt->execute();
+
+        $hotels = $stmt->fetchAll(PDO::FETCH_OBJ);
+
+        $db = null;
+
+        return $response->write(json_encode($hotels))->withStatus(200);
+    } catch (PDOException $e) {
+        $db = null;
+
+        $errorMessage = $e->getMessage();
+        return $response->write(json_encode(['error' => '' . $errorMessage]))->withStatus(500);
+    }
+});
+
 $app->get('/api/hotels/{id}', function (Request $request, Response $response, array $args) {
 
```

Comment:
For the search hotel by name and branch, we’d better use column1 LIKE '%this%'. Otherwise, it is hard for the user to find a hotel.

Reply:
This is not part of this PR, and this endpoint is not for search. This is getting the hotel by the id. The search feature is not part of this PR

Reply:
Yeah, but that is a problem on master branch

Reply:
Then we should create a new branch for solving that problem. It should not be in this branch because this PR is about availability. 
Also the `like` in sql is not very intelligence, we need to talk about this later about how we wanna to the blur search

Reply:
We'd better think it again. Overall, this PR looks good to me.

Reply:
@klQiu @suiyoubi  good job for this discussion.